### ISSUE TITLE
Rule proposal: `prefer-await`

### ISSUE THREAD

**Issue by yvele** [sindresorhus/eslint-plugin-unicorn#652]
The rule name may be named `unicorn/prefer-await`.

This only applies for ES6 code that have the [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) keyword available.

#  Prefer `await` over `Promise#then()`, `Promise#catch()` and `Promise#finally()`

Prefer using [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) operator over [`Promise#then()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then), [`Promise#catch()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) and [`Promise#finally()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally), which is easier to write and to read afterwards.

~~This rule is fixable.~~

## Fail

```js
await doSomething().then(() => {
  // …
});
```

```js
doSomething().then(() => {
  // …
});
```

```js
await doSomething().catch((err) => {
  // …
});
```

```js
doSomething().catch((err) => {
  // …
});
```

## Pass

```js
 await doSomething();
```

```js
try {
  await doSomething();
} catch (err) {
  // …
}
```

**Comment by fisker** [comment]
How about `no-promise-catch`.

**Comment by yvele** [comment]
Yeah right `prefer-await` fits better and will forbid using `.then()`, `.catch()` and `.finally()` methods.

**Comment by fisker** [comment]
There is one rule called [`prefer-await-to-then`](https://github.com/xjamundx/eslint-plugin-promise/blob/master/rules/prefer-await-to-then.js).

There is also proposal in eslint project, https://github.com/eslint/eslint/issues/9649 and also a PR https://github.com/eslint/eslint/pull/11962 :smile:

**Comment by yvele** [comment]
> There is one rule called [`prefer-await-to-then`](https://github.com/xjamundx/eslint-plugin-promise/blob/master/rules/prefer-await-to-then.js) [](/xjamundx/eslint-plugin-promise/blob/HEAD@{2020-03-31T12:12:22Z}/rules/prefer-await-to-then.js).

Sure but the repo looks a bit abandoned.

There is also sommes issues with this rule 😅 like [this one](https://github.com/xjamundx/eslint-plugin-promise/issues/179). Or that fact that it [only catches `.then()` and not `.catch()` used alone](https://github.com/xjamundx/eslint-plugin-promise/blob/5b935bdd3c3760a2e58eea9b89c86b6d3243e321/rules/prefer-await-to-then.js#L42). Also the rule is not fixable.

> There is also proposal in eslint project, [eslint/eslint#9649](https://github.com/eslint/eslint/issues/9649) and also a PR [eslint/eslint#11962](https://github.com/eslint/eslint/pull/11962) 😄

Looks like they abandoned this rule in to be integrated in core ESLint rule [for performance control reasons](https://github.com/eslint/eslint/pull/11961#issuecomment-536333393) 🤔 

So do you want me to close this issue in favor of `xjamundx/eslint-plugin-promise`? 🤔

**Comment by sindresorhus** [comment]
I'm :+1: on adding a `prefer-await` rule here. @fisker ?

**Comment by fisker** [comment]
I like this, remember I was looking for similar rule before. We need discuss details.

1. Do we supply auto fix?  I hate rewrite those promises.

2. If not inside asyc functions, how do we handle?We can't change function to async, since old function could return both `Promise` and other types.

3. top level `await`? Behind option?

**Comment by sindresorhus** [comment]
1. We could do so for the simple cases, but it's not important to me as all my code is already using only async/await and I only need this rule to enforce so for new contributions.
2. We cannot fix then, but we should still report. We could maybe add a suggestion (suggestions API) to add the `async` keyword and then fix, but not sure it's worth the effort.
3. Nah, we can open an issue about it and implement it when Node.js stable supports top-level `await`.

**Comment by fisker** [comment]
I tried a little bit to implement this rule, starting on fixing `.then`, really complicated...

**Comment by fisker** [comment]
```js
promise.then(onFulfilled, onRejected);
```

Shouldn't fix to

```js
try {
  onFulfilled(await promise);
} catch (error) {
  onRejected(error);
}
```

`onRejected` should not handle `onFulfilled` errors, right?


So, the following one is right?

```js
let result;
try {
  result = await promise;
} catch (error) {
  onRejected(error);
}
onFulfilled(result)
```

**Comment by yvele** [comment]
@fisker maybe the rule should not fix that kind of case (that requires a new variable).. and the rule should be partially fixable.

Note that this is not an acceptable fix because we don't want to catch `onFulfilled`:

```js
try {
  onFulfilled(await promise);
} catch (error) {
  onRejected(error);
}
```

**Comment by fisker** [comment]
I have lots of legacy codes need rewrite from my work, I want a fix so much.

**Comment by yvele** [comment]
Yeah but what if a `result` variable already exists on the same scope? Will this use case be the limit of what is fixable?

**Comment by fregante** [comment]
This exists in `eslint-plugin-promise` and works well!

https://github.com/xjamundx/eslint-plugin-promise/blob/development/docs/rules/prefer-await-to-then.md

**Comment by fisker** [comment]
Reopening this, since `eslint-plugin-promise` don't have autofix. I'm going to give another shot on it.

**Comment by papb** [comment]
Hi @fisker, you might be interested in https://github.com/codemodsquad/asyncify by @jedwards1211 - it was used to convert the whole codebase of https://github.com/sequelize/sequelize from Bluebird promises to async-await.

**Comment by fisker** [comment]
Thanks, I'll take a look.

**Comment by jedwards1211** [comment]
It would probably be the most complicated autofix ever 🤣
Even my codemod has limitations, esp if there's a lot of branching inside promise handlers.

**Comment by jedwards1211** [comment]
Also I never implemented support for Flow or TS type annotations in my codemod, which would probably be a minimum requirement for an eslint autofix.  Would probably take a lot more work to fully support them

**Comment by fisker** [comment]
Thanks for you information, `asyncify` looks great, do you think it's possible to covert your work to ESLint rule?

**Comment by jedwards1211** [comment]
Hmmm, actually type annotations aren't a huge problem, I just fixed an issue with catch error type annotations, but the argument type annotations already get transferred by `asyncify` during unrolling...

If you mean convert to an ESLint autofix, I'm sure it's possible, but the main obstacle is `asyncify` uses Babel API everywhere, which isn't the default eslint parser.  I have no idea how eslint rules and autofixes are written to work with a variety of parsers.  I'm happy to keep fixing issues with `asyncify` but I wouldn't have time to convert it to an eslint autofix...IMO it's better not to have an autofix for this and rely on an external tool like `asyncify` to autofix, since autofixing is a very involved and fraught process.

Plus, people might have preferences for what the autofix does.  For example, not everyone would want to unroll `return foo.catch(() => null)` to

```js
try {
  return await foo
} catch (err) {
  return null
}
```
So I'm not sure at autofix would satisfy everyone.  `asyncify` already has an `ignoreChainsShorterThan` option.  I guess that could be an eslint option too, but I imagine the lint check would have to support options like that as well.

**Comment by jedwards1211** [comment]
Another consideration: a lot of code has to get moved around in an autofix.  I used `recast` to try to preserve as much formatting as possible on code that gets moved.  I assume an eslint autofix would preserve less formatting, but I don't really know.

**Comment by fisker** [comment]
Thanks a lot.

**Comment by jedwards1211** [comment]
Oh and @papb thanks for the shout-out!

**Comment by papb** [comment]
@jedwards1211 And thank you for all that hard work 😁

**Comment by yvele** [comment]
> For example, not everyone would want to unroll oneliners like `return foo.catch(() => null)` to five whole lines

I personally prefer the 5 whole lines as I want developers to catch specific errors and also add comments about why the error needs to be catched.

```js
try {
  return await foo
} catch (err) {
  if (err.name === "bar") {
    // Error is swallowed because blabla.
    return null;
  }
  throw err;
}
```

In my opinion code like `return foo.catch(() => null)` reflects more conceptual issues then styling issues.

**Comment by jedwards1211** [comment]
@yvele even if you do handling like that in a catch callback, I still prefer the catch callback sometimes, because try/catch is so ungainly to use.  Especially if you have to hoist a variable declaration outside of the try block, it's pretty terrible.

**Comment by jedwards1211** [comment]
With `.catch` you can at least keep a clean variable declaration:

```js
const result: Foo | null = await foo.catch((err: Error) => {
  if (err.name === "bar") {
    // Error is swallowed because blabla.
    return null;
  }
  throw err;
})
```

**Comment by fregante** [comment]
I agree that `await foo.catch(x => undefined)` is a great way to ignore errors or reassign them to something non-throwing. `try/catch` can be so noisy for that

```js
let result: Foo;
try {
  result = await foo;
} catch(err: unknown) {
  if (err.name === "bar") {
    // Error is swallowed because blabla.
  } else {
    throw err;
  }
}
```
